annotation-processor/src/main/java/org/hibernate/validator/ap/classchecks/AbstractMethodOverrideCheck.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.ap.classchecks;
+
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+import org.hibernate.validator.ap.checks.ConstraintCheckIssue;
+import org.hibernate.validator.ap.util.CollectionHelper;
+
+/**
+ * Abstract base class for {@link ClassCheck} implementations that check overridden methods.
+ *
+ * @author Marko Bekhta
+ */
+public abstract class AbstractMethodOverrideCheck extends AbstractClassCheck {
+
+	private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+	private static final String JAVAX_VALIDATION_CONSTRAINT = "javax.validation.Constraint";
+
+	private final Elements elementUtils;
+
+	private final Types typeUtils;
+
+	public AbstractMethodOverrideCheck(Elements elementUtils, Types typeUtils) {
+		this.elementUtils = elementUtils;
+		this.typeUtils = typeUtils;
+	}
+
+	@Override
+	public Collection<ConstraintCheckIssue> checkMethod(ExecutableElement currentMethod) {
+		if ( !needToPerformAnyChecks( currentMethod ) ) {
+			return Collections.emptySet();
+		}
+
+		TypeElement currentTypeElement = getEnclosingTypeElement( currentMethod );
+
+		// if current type is java.lang.Object then we can continue without doing any other checks
+		if ( isJavaLangObjectOrNull( currentTypeElement ) ) {
+			return Collections.emptySet();
+		}
+		// find if there's a method that was overridden by the current one.
+		Map<Boolean, List<ExecutableElement>> overriddenMethods = findAllOverriddenElements( currentTypeElement, currentMethod );
+		if ( overriddenMethods.get( Boolean.TRUE ).isEmpty() && overriddenMethods.get( Boolean.FALSE ).isEmpty() ) {
+			return Collections.emptySet();
+		}
+
+		// if there's more than one overridden originally declared method we need to make sure that all of them match
+		Set<ConstraintCheckIssue> errors = CollectionHelper.newHashSet();
+		for ( ExecutableElement firstExecutable : overriddenMethods.get( Boolean.TRUE ) ) {
```

Comment:
if there's noting special for `@Valid` then everything else implemented. I've removed all the check logic from the `checkMethod` in abstract class. What I left there is a check to see if any further checks need to be made (call to abstract `AbstractMethodOverrideCheck#needToPerformAnyChecks`) and getting the overridden methods (if there's no overridden methods then no need to check anything else) as these two things are common. After that a call to abstract `AbstractMethodOverrideCheck#checkMethodInternal` is made that do the real check of parameters/return value.

I've also cleaned up the code and used `ConstraintHelper` methods to determine if it's a constraint annotation or `@Valid` (previously there was a set of private methods in `AbstractMethodOverrideCheck` to do this kind of filtering of annotations)

Please let me know if anything else is needed. And thanks again for your patience and time spent on this ticket !